pandas/tests/groupby/test_transform.py

```diff
@@ -317,6 +317,32 @@ def test_dispatch_transform(tsframe):
     tm.assert_frame_equal(filled, expected)
 
 
+def test_transform_transformation_func(transformation_func):
+    # GH 30918
+    df = DataFrame(
+        {
+            "A": ["foo", "foo", "foo", "foo", "bar", "bar", "baz"],
+            "B": [1, 2, np.nan, 3, 3, np.nan, 4],
+        }
+    )
+
+    if transformation_func in ["pad", "backfill", "tshift", "corrwith", "cumcount"]:
+        # These transformation functions are not yet covered in this test
+        return
+    elif transformation_func == "fillna":
+        test_op = lambda x: x.transform("fillna", value=0)
+        mock_op = lambda x: x.fillna(value=0)
+    else:
+        test_op = lambda x: x.transform(transformation_func)
+        mock_op = lambda x: getattr(x, transformation_func)()
+
+    result = test_op(df.groupby("A"))
+    groups = [df[["B"]].iloc[:4], df[["B"]].iloc[4:6], df[["B"]].iloc[6:]]
```

Comment:
I think instead of concat-ing like this would be easier if you did something like:

```python
grp = df.groupby("A")
result = test_op(grp)
expected = mock_op(grp)
```

But let me know if overlooking something

Reply:
It's easier, but I just thought this removes the dependency on implementation of groupby functions (e.g. `groupby.fillna`, `groupby.ffill`), which may share the same root with what we are testing (e.g. `groupby.transform('fillna', ...)`, `groupby.transform('ffill')`).

Using `concat` on the other hand is clean and explicit.

Reply:
Gotcha. Ok for now though might be nice in a follow up to extend this test to reduction functions as well, which I think would require this

Reply:
Sure. There are quite a few things we need to fix / cleanup here, such as moving corrwith from transformation_kernels to reduction_kernels referenced in xfails. I will fix these one by one, and create this parallel test for reduction functions.